### PR TITLE
Logs Tab: Show log line count

### PR DESCRIPTION
**Pull request by gtk-grafana**
Fixes: https://github.com/grafana/explore-logs/issues/425

image

image

image

image

\@TODO:
* ✅  we're running this query on every activation of the ServiceScene, can we only run it when the query/timerange changes or if the current count is undefined?
* ✅  fix e2e assertions now that there's another query in the mix
* ✅  LogsCount needs to sum all infinite scroll pages
* ✅ Custom maxLines support

### CODE REVIEW

File: src/Components/ServiceScene/ActionBarScene.tsx [hunk 1]
```diff
@@ -168,3 +175,55 @@ function ToolbarExtensionsRenderer(props: { serviceScene: SceneObject }) {
...
+
+function LogsCount(className: string | undefined, logsCount: number | undefined, totalCount: number | undefined) {
+  const styles = useStyles2(getLogsCountStyles);
+
+  if (logsCount !== undefined) {
+    const valueFormatter = getValueFormat('short');
+
+    const formattedLogsCount = valueFormatter(logsCount, 0);
+
+    if (logsCount < LINE_LIMIT || totalCount === undefined) {
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/ActionBarScene.tsx:187]: Hmm, will this get messed up by infinite scrolling?

Comment (gtk-grafana) [reply src/Components/ServiceScene/ActionBarScene.tsx:187]: Not this, but the logsCount total does not sum multiple pages

**Comment by gtk-grafana** [comment]
@matyax There doesn't seem to be any way for the app to know when a new page has been loaded in the LogsPanel with infinite scroll, the observable is not updated, so the panel never goes into the loading state, and there's nothing for us to subscribe to from the app.

Do we need to emit an event from the panel plugin?

**Comment by matyax** [comment]
Would maybe this help? https://github.com/grafana/explore-logs/blob/main/src/Components/ServiceScene/LogsPanelScene.tsx#L128

File: src/services/query.ts [hunk 2]
```diff
@@ -225,3 +228,5 @@ export function unwrapWildcardSearch(input: string) {
 export function sanitizeStreamSelector(expression: string) {
   return expression.replace(/\s*,\s*}/, '}');
 }
+
+export const LINE_LIMIT = 1000;
```
Comment (svennergr) [inline src/services/query.ts:232]:
```suggestion
// default line limit; each data source can define it's own line limit too
export const LINE_LIMIT = 1000;
```

File: src/Components/ServiceScene/ServiceScene.tsx [hunk 3]
```diff
@@ -397,8 +430,29 @@ export class ServiceScene extends SceneObjectBase<ServiceSceneState> {
   }
 
   private subscribeToLogsQuery() {
-    return this.state.$data?.subscribeToState((newState) => {
+    return this.state.$data?.subscribeToState((newState, prevState) => {
       this.updateLoadingState(newState, TabNames.logs);
+      if (newState.data?.state === LoadingState.Done || newState.data?.state === LoadingState.Streaming) {
+        const resultCount = newState.data.series[0].length;
```
Comment (svennergr) [inline src/Components/ServiceScene/ServiceScene.tsx:436]:
```suggestion
        const resultCount = newState.data.series[0]?.length ?? 0;
```
Just as a safeguard if there's no series?

File: src/Components/ServiceScene/ServiceScene.tsx [hunk 4]
```diff
@@ -397,8 +430,29 @@ export class ServiceScene extends SceneObjectBase<ServiceSceneState> {
...
+        }
+      }
+    });
+  }
+
+  private subscribeToLogsCountQuery() {
+    return this.state.$logsCount?.subscribeToState((newState) => {
+      if (newState.data?.state === LoadingState.Done) {
+        const value: number | undefined = newState.data.series[0]?.fields?.[1]?.values?.[0];
+        if (value !== undefined) {
```
Comment (svennergr) [inline src/Components/ServiceScene/ServiceScene.tsx:450]: What happens if it is undefined and the `totalLogsCount` state is currently set? Should we reset to 0?

**Review by svennergr — APPROVED** [review]
Left a few nits. For my own taste, we could spare the displayed log count and just show the number of total logs, but that's on you and @zizzpudding, I'd say.

File: src/Components/ServiceScene/ServiceScene.tsx (hunk 4 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/ServiceScene.tsx]: If the instant query returns "Done" but does not return a result, I guess that would have to be a no-data state, so yeah we should set 0, but the total count won't get displayed if the logs query returns 0, so hopefully it doesn't matter either way.

Updated!

**Comment by gtk-grafana** [comment]
@svennergr I chatted with @zizzpudding, we decided to move the log line count to the logs/table panel header, and additionally place the total log line count in the log volume panel header.

**Comment by gtk-grafana** [comment]
@svennergr I updated the UI, however in my testing I discovered a routing bug that was causing labels containing `/` to redirect back to the same route on every label update, but more importantly I realized how "off" the instant query results are compared to the log volume query, particularly for low volume/ small duration queries.

image

So I added the check back in there to output the results of the logs panel query if the instant query is undefined, and the logs panel returned less then the line count. 

image

**Comment by gtk-grafana** [comment]
Oops we forgot the table header

**Comment by gtk-grafana** [comment]
@svennergr I added the count to the table as well.
@matyax I found a bug with the data frame: https://github.com/grafana/explore-logs/issues/966

Also we might want to remove the logs count in the footer now, but let's do that in another PR.

**Pull request merged**